core/src/main/java/org/mskcc/cbio/portal/dao/DaoReferenceGenomeGene.java

```diff
@@ -0,0 +1,255 @@
+                pstmt.setLong(1, gene.getEntrezGeneId());
+                pstmt.setInt(2, gene.getReferenceGenomeId());
+                pstmt.setString(3, gene.getChr());
+                pstmt.setString(4, gene.getCytoband());
+                pstmt.setInt(5, gene.getExonicLength());
+                pstmt.setLong(6, gene.getStart());
+                pstmt.setLong(7, gene.getEnd());
+                rows += pstmt.executeUpdate();
+            }
+            else {
+                rows += updateGene(existingGene);
+            }
+            return rows;
+        } catch (SQLException e) {
+            throw new DaoException(e);
+        } finally {
+            JdbcUtil.closeAll(DaoGene.class, con, pstmt, rs);
+        }
+    }
+
+
+    /**
+     * Gets the Gene with the Specified Entrez Gene ID.
+     * For faster access, consider using DaoGeneOptimized.
+     *
+     * @param entrezGeneId ENTRZ GENE ID.
+     * @return Canonical Gene Object.
+     * @throws DaoException Database Error.
+     */
+    public ReferenceGenomeGene getGene(long entrezGeneId, int referenceGenomeId) throws DaoException {
+        Connection con = null;
+        PreparedStatement pstmt = null;
+        ResultSet rs = null;
+        try {
+            con = JdbcUtil.getDbConnection(DaoReferenceGenomeGene.class);
+            pstmt = con.prepareStatement
+                ("SELECT * FROM `reference_genome_gene` WHERE `ENTREZ_GENE_ID` = ? AND `REFERENCE_GENOME_ID` = ?");
+            pstmt.setLong(1, entrezGeneId);
+            pstmt.setInt(2, referenceGenomeId);
+            rs = pstmt.executeQuery();
+            if (rs.next()) {
+                return extractGene(rs);
+            } else {
+                return null;
+            }
+        } catch (SQLException e) {
+            throw new DaoException(e);
+        } finally {
+            JdbcUtil.closeAll(DaoGene.class, con, pstmt, rs);
+        }
+    }
+
+    /**
+     * Gets the Gene with the Specified Entrez Gene ID.
+     * For faster access, consider using DaoGeneOptimized.
+     *
+     * @param hugoGeneSymbol Hugo Gene Symbol.
+     * @return Canonical Gene Object.
+     * @throws DaoException Database Error.
+     */
+    public ReferenceGenomeGene getGene(String hugoGeneSymbol, int referenceGenomeId) throws DaoException {
+        Connection con = null;
+        PreparedStatement pstmt = null;
+        ResultSet rs = null;
+        try {
+            con = JdbcUtil.getDbConnection(DaoReferenceGenomeGene.class);
+            pstmt = con.prepareStatement
+                ("SELECT * FROM `reference_genome_gene` JOIN `gene` ON `reference_genome_gene`.entrez_gene_id=" +
+                    "`gene`.entrez_gene_id WHERE `HUGO_GENE_SYMBOL` = ? AND `REFERENCE_GENOME_ID` = ?");
+            pstmt.setString(1, hugoGeneSymbol);
+            pstmt.setInt(2, referenceGenomeId);
+            rs = pstmt.executeQuery();
+            if (rs.next()) {
+                return extractGene(rs);
```

Comment:
we use the same extractGene() function here and also in the previous function call which does not join tables. I guess I wonder: are we getting any additional information from the gene table when extracting this gene? If not, then why are we joining the tables?

Ah .. I think I see --- this function takes hugo gene symbol, and we are using that as a search criteria against the gene table. Maybe a more natural way to do this would be to first search the gene table with hugo gene symbol and resolve it to an entrez gene id. After that, you can call the prior function with the entrez_gene_id and the reference_genome_id.

Another issue here is that we use the gene_alias table sometimes to resolve hugo gene symbols. The way this is written, getGene(String hugoGeneSymbol, int referenceGenomeId) is only going to find exact matches to Hugo gene symbols in the gene table. Do we not need to resolve based on aliases sometimes?

Reply:
no so sure if we any use case like querying based on gene aliases. 
ah, good idea. We could make two calls: call to the gene table for retrieving hugoGeneSymbol followed by the one to the reference_genome_gene table. It might be faster this way since both tables are cached.